File: src/StockAdvisorFS/WorkflowSample.fs
module StockAdvisorFS.WorkflowSample

open System
open System.Threading.Tasks
open AgentNet
open Anthropic
open Microsoft.Extensions.AI

// ============================================================================
// WORKFLOW SAMPLE
// Demonstrates the AgentNet workflow DSL with parallel execution,
// sequential pipeline, and AI agent integration for comparing two stocks.
// ============================================================================

// ----------------------------------------------------------------------------
// Chat Client Setup
// ----------------------------------------------------------------------------

let private failIfNone msg opt = opt |> Option.defaultWith (fun () -> failwith msg)
let private tryGetEnv = Environment.GetEnvironmentVariable >> Option.ofObj

let private createChatClient () =
    let apiKey = tryGetEnv "ANTHROPIC_API_KEY" |> failIfNone "ANTHROPIC_API_KEY environment variable is not set."
    let model = tryGetEnv "ANTHROPIC_MODEL" |> Option.defaultValue "claude-sonnet-4-20250514"
    let client = AnthropicClient(APIKey = apiKey)
    client.AsIChatClient(model)

let private createAnalysisAgent (chatClient: IChatClient) =
    Agent.create """
        You are a stock comparison analyst. Given information about two stocks,
        provide a concise comparison highlighting:
        - Which stock appears stronger and why
        - Key differences in their metrics
        - Risk considerations
        - A brief recommendation
        Keep your response focused and under 200 words.
        """
    |> Agent.withName "StockAnalyst"
    |> Agent.build chatClient

// ----------------------------------------------------------------------------
// Domain Types
// ----------------------------------------------------------------------------

type StockSymbols = { Symbol1: string; Symbol2: string }
type StockData = { Symbol: string; Info: string; Volatility: string }
type StockPair = { Stock1: StockData; Stock2: StockData }
type AnalysisResult = { Pair: StockPair; Analysis: string }

// ----------------------------------------------------------------------------
// Workflow Executors
// ----------------------------------------------------------------------------

/// Fetches stock info and volatility for a symbol
let private fetchStockData (symbol: string) : Task<StockData> = task {
    let! info = StockTools.getStockInfo symbol
    let! volatility = StockTools.calculateVolatility symbol
    return { Symbol = symbol; Info = info; Volatility = volatility }
}

/// Fetches data for both stocks in parallel
let private fetchBothStocks (symbols: StockSymbols) : Task<StockPair> = task {
    let! results = Task.WhenAll([|
        fetchStockData symbols.Symbol1
        fetchStockData symbols.Symbol2
    |])
    return { Stock1 = results.[0]; Stock2 = results.[1] }
}

/// Creates an executor that uses the AI agent to compare stocks
let private createCompareExecutor (agent: ChatAgent) =
    let formatPrompt (pair: StockPair) =
        $"""Compare these two stocks:

{pair.Stock1.Symbol}:
{pair.Stock1.Info}
Volatility: {pair.Stock1.Volatility}

{pair.Stock2.Symbol}:
{pair.Stock2.Info}
Volatility: {pair.Stock2.Volatility}"""

    let parseResponse pair response =
        { Pair = pair; Analysis = response }

    Executor.fromAgentWith "CompareStocks" formatPrompt parseResponse agent

/// Generates a formatted report
let private generateReport (result: AnalysisResult) : string =
    $"""
================================================================================
                        STOCK COMPARISON REPORT
================================================================================

--- {result.Pair.Stock1.Symbol} ---
{result.Pair.Stock1.Info}

{result.Pair.Stock1.Volatility}

--- {result.Pair.Stock2.Symbol} ---
{result.Pair.Stock2.Info}

{result.Pair.Stock2.Volatility}

--- AI ANALYSIS ---
{result.Analysis}
================================================================================
"""

// ----------------------------------------------------------------------------
// Workflow Definition
// ----------------------------------------------------------------------------

/// Multi-stock comparison workflow using the AgentNet workflow DSL
let private stockComparisonWorkflow (agent: ChatAgent) = workflow {
    start (Executor.fromTask "FetchBothStocks" fetchBothStocks)
    next (createCompareExecutor agent)
    next (Executor.fromFn "GenerateReport" generateReport)
}

// ----------------------------------------------------------------------------
// Public Entry Point
// ----------------------------------------------------------------------------

/// Runs the workflow sample comparing two stocks
let run (symbol1: string) (symbol2: string) : Task<unit> = task {
    printfn $"\nRunning comparison workflow for {symbol1} vs {symbol2}..."
    printfn "Initializing AI agent..."

    let chatClient = createChatClient ()
    let agent = createAnalysisAgent chatClient
    let wf = stockComparisonWorkflow agent

    printfn "Step 1: Fetching stock data in parallel..."
    printfn "Step 2: AI agent analyzing stocks..."

    let input = { Symbol1 = symbol1; Symbol2 = symbol2 }
    let! report = Workflow.run input wf

    printfn "%s" report
}

File: src/StockAdvisorCS/WorkflowSample.cs
using Anthropic;
using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;

namespace StockAdvisorCS;

// ============================================================================
// WORKFLOW SAMPLE
// Demonstrates the MAF WorkflowBuilder with parallel execution,
// sequential pipeline, and AI agent integration for comparing two stocks.
// ============================================================================

// ----------------------------------------------------------------------------
// Domain Types
// ----------------------------------------------------------------------------

public record StockSymbols(string Symbol1, string Symbol2);
public record StockData(string Symbol, string Info, string Volatility);
public record StockPair(StockData Stock1, StockData Stock2);
public record AnalysisResult(StockPair Pair, string Analysis);

// ----------------------------------------------------------------------------
// Workflow Executors
// ----------------------------------------------------------------------------

internal sealed class FetchBothStocksExecutor() : Executor<StockSymbols, StockPair>("FetchBothStocks")
{
    public override async ValueTask<StockPair> HandleAsync(
        StockSymbols symbols,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        Console.WriteLine("  [FetchBothStocks] Fetching stock data in parallel...");

        var task1 = FetchStockDataAsync(symbols.Symbol1);
        var task2 = FetchStockDataAsync(symbols.Symbol2);
        await Task.WhenAll(task1, task2);

        return new StockPair(task1.Result, task2.Result);
    }

    private static async Task<StockData> FetchStockDataAsync(string symbol)
    {
        var info = await StockTools.GetStockInfo(symbol);
        var volatility = await StockTools.CalculateVolatility(symbol);
        return new StockData(symbol, info, volatility);
    }
}

internal sealed class CompareStocksExecutor(AIAgent agent) : Executor<StockPair, AnalysisResult>("CompareStocks")
{
    public override async ValueTask<AnalysisResult> HandleAsync(
        StockPair pair,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        Console.WriteLine("  [CompareStocks] AI agent analyzing stocks...");

        var prompt = $"""
            Compare these two stocks:

            {pair.Stock1.Symbol}:
            {pair.Stock1.Info}
            Volatility: {pair.Stock1.Volatility}

            {pair.Stock2.Symbol}:
            {pair.Stock2.Info}
            Volatility: {pair.Stock2.Volatility}
            """;

        var thread = agent.GetNewThread();
        var response = await agent.RunAsync(prompt, thread);
        return new AnalysisResult(pair, response.Text);
    }
}

internal sealed class GenerateReportExecutor() : Executor<AnalysisResult, string>("GenerateReport")
{
    public override ValueTask<string> HandleAsync(
        AnalysisResult result,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        Console.WriteLine("  [GenerateReport] Generating report...");
        var report = $"""

================================================================================
                        STOCK COMPARISON REPORT
================================================================================

--- {result.Pair.Stock1.Symbol} ---
{result.Pair.Stock1.Info}

{result.Pair.Stock1.Volatility}

--- {result.Pair.Stock2.Symbol} ---
{result.Pair.Stock2.Info}

{result.Pair.Stock2.Volatility}

--- AI ANALYSIS ---
{result.Analysis}
================================================================================
""";
        return ValueTask.FromResult(report);
    }
}

// ----------------------------------------------------------------------------
// Workflow Sample
// ----------------------------------------------------------------------------

public static class WorkflowSample
{
    private static IChatClient CreateChatClient()
    {
        var apiKey = Environment.GetEnvironmentVariable("ANTHROPIC_API_KEY")
            ?? throw new InvalidOperationException("ANTHROPIC_API_KEY environment variable is not set.");
        var model = Environment.GetEnvironmentVariable("ANTHROPIC_MODEL") ?? "claude-sonnet-4-20250514";

        var client = new AnthropicClient { ApiKey = apiKey };
        return client.AsIChatClient(model);
    }

    private static AIAgent CreateAnalysisAgent(IChatClient chatClient)
    {
        return new ChatClientAgent(
            chatClient,
            name: "StockAnalyst",
            instructions: """
                You are a stock comparison analyst. Given information about two stocks,
                provide a concise comparison highlighting:
                - Which stock appears stronger and why
                - Key differences in their metrics
                - Risk considerations
                - A brief recommendation
                Keep your response focused and under 200 words.
                """);
    }

    private static Workflow BuildWorkflow(AIAgent agent)
    {
        // Create executors
        var fetchBothStocks = new FetchBothStocksExecutor();
        var compareStockPair = new CompareStocksExecutor(agent);
        var reportExecutor = new GenerateReportExecutor();

        // Build the workflow using MAF WorkflowBuilder
        WorkflowBuilder builder = new(fetchBothStocks);
        builder.AddEdge(fetchBothStocks, compareStockPair);
        builder.AddEdge(compareStockPair, reportExecutor);
        builder.WithOutputFrom(reportExecutor);
        return builder.Build();
    }

    /// <summary>
    /// Runs the workflow sample comparing two stocks.
    /// </summary>
    public static async Task RunAsync(string symbol1, string symbol2)
    {
        Console.WriteLine($"\nRunning MAF workflow for {symbol1} vs {symbol2}...");
        Console.WriteLine("Initializing AI agent...");

        var chatClient = CreateChatClient();
        var agent = CreateAnalysisAgent(chatClient);
        var workflow = BuildWorkflow(agent);

        var input = new StockSymbols(symbol1, symbol2);

        await using var run = await InProcessExecution.RunAsync(workflow, input);

        // Process workflow events
        foreach (var evt in run.NewEvents)
        {
            if (evt is ExecutorCompletedEvent completed && completed.ExecutorId == "GenerateReport")
            {
                Console.WriteLine(completed.Data);
            }
        }
    }
}
